universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -31,11 +36,18 @@ public ProgramServiceImpl(ProgramRepository programRepository) {
   public CompletionStage<ImmutableList<ProgramDefinition>> listProgramDefinitionsAsync() {
     return programRepository
         .listPrograms()
-        .thenApply(
+        .thenComposeAsync(
             programs ->
-                programs.stream()
-                    .map(program -> program.getProgramDefinition())
-                    .collect(ImmutableList.toImmutableList()));
+                questionService
+                    .getReadOnlyQuestionService()
```

Comment:
suggestion (blocking): `getReadOnlyQuestionService` and `listPrograms can be done in parallel rather than series. Something like:

```java
// method body
CompleteableFuture<ReadOnlyQuestionService> roQuestionServiceFuture = questionService.getReadOnlyQuestionService().toCompleteableFuture();
CompleteableFuture<ImmutableList<Program>> programsFuture = programRepository.listPrograms().toCompleteableFuture();

return CompleteableFuture.allOf(roQuestionServiceFuture, programsFuture).thenApplyAsync(
    () -> {
      ReadOnlyQuestionService roQuestionService = roQuestionServiceFuture.join();
      ImmutableList<Program> programs = programsFuture.join();
    
      return programs.stream().map(...);
    }, 
  // please specify the default threadpool explicitly
  // see https://www.playframework.com/documentation/2.8.x/ThreadPools#Using-the-default-thread-pool
  // for more details 
  httpExecutionContext.current);

```

Reply:
Done